lib/src/barback/dartdevc/dartdevc.dart

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:io';
+
+import 'package:barback/barback.dart';
+import 'package:bazel_worker/bazel_worker.dart';
+import 'package:cli_util/cli_util.dart' as cli_util;
+import 'package:path/path.dart' as p;
+
+import '../../io.dart';
+import 'module.dart';
+import 'module_reader.dart';
+import 'scratch_space.dart';
+import 'summaries.dart';
+import 'workers.dart';
+
+/// Bootstraps the js module for the entrypoint dart file [dartEntrypointId]
+/// with two additional JS files:
+///
+/// * A `$dartEntrypointId.js` file which is the main entrypoint for the app. It
+///   injects a script tag whose src is `require.js` and whose `data-main`
+///   attribute points at a `$dartEntrypointId.bootstrap.js` file.
+/// * A `$dartEntrypointId.bootstrap.js` file which invokes the top level `main`
+///   function from the entrypoint module, after performing some necessary SDK
+///   setup.
+///
+/// In debug mode an empty sourcemap will be output for the entrypoint js file
+/// to satisfy the test package runner (there is no original dart file to map it
+/// back to though).
+///
+/// Synchronously returns a `Map<AssetId, Future<Asset>>` so that you can know
+/// immediately what assets will be output.
+Map<AssetId, Future<Asset>> bootstrapDartDevcEntrypoint(
+    AssetId dartEntrypointId, BarbackMode mode, ModuleReader moduleReader) {
+  var bootstrapId = dartEntrypointId.addExtension('.bootstrap.js');
+  var jsEntrypointId = dartEntrypointId.addExtension('.js');
+  var jsMapEntrypointId = jsEntrypointId.addExtension('.map');
+
+  var outputCompleters = <AssetId, Completer<Asset>>{
+    bootstrapId: new Completer<Asset>(),
+    jsEntrypointId: new Completer<Asset>(),
+  };
+  if (mode == BarbackMode.DEBUG) {
+    outputCompleters[jsMapEntrypointId] = new Completer<Asset>();
+  }
+
+  () async {
+    var module = await moduleReader.moduleFor(dartEntrypointId);
+
+    // The path to the entrypoint js module as it should appear in the call to
+    // `require` in the bootstrap file.
+    var moduleDir = topLevelDir(dartEntrypointId.path);
+    var appModulePath = p.relative(p.join(moduleDir, module.id.name),
+        from: p.dirname(dartEntrypointId.path));
```

Comment:
Are these local file paths or URIs? Should these be `p.url` here? What about below?